ports/esp32/machine_adc.c

```diff
@@ -31,178 +32,414 @@
+    {{&machine_adc_type}, ADCBLOCK1, ADC_CHANNEL_3, GPIO_NUM_4, -1},
+    {{&machine_adc_type}, ADCBLOCK1, ADC_CHANNEL_4, GPIO_NUM_5, -1},
+    {{&machine_adc_type}, ADCBLOCK1, ADC_CHANNEL_5, GPIO_NUM_6, -1},
+    {{&machine_adc_type}, ADCBLOCK1, ADC_CHANNEL_6, GPIO_NUM_7, -1},
+    {{&machine_adc_type}, ADCBLOCK1, ADC_CHANNEL_7, GPIO_NUM_8, -1},
+    {{&machine_adc_type}, ADCBLOCK1, ADC_CHANNEL_8, GPIO_NUM_9, -1},
+    {{&machine_adc_type}, ADCBLOCK1, ADC_CHANNEL_9, GPIO_NUM_10, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_0, GPIO_NUM_11, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_1, GPIO_NUM_12, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_2, GPIO_NUM_13, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_3, GPIO_NUM_14, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_4, GPIO_NUM_15, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_5, GPIO_NUM_16, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_6, GPIO_NUM_17, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_7, GPIO_NUM_18, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_8, GPIO_NUM_19, -1},
+    {{&machine_adc_type}, ADCBLOCK2, ADC_CHANNEL_9, GPIO_NUM_20, -1},
     #endif
 };
 
-STATIC uint8_t adc_bit_width;
 
-STATIC mp_obj_t madc_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw,
-    const mp_obj_t *args) {
+STATIC void madcblock_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    madcblock_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "ADCBlock(%u, bits=%u)", self->unit_id, self->bits);
+}
+
 
-    static int initialized = 0;
-    if (!initialized) {
+STATIC void madcblock_bits_helper(madcblock_obj_t *self, mp_int_t bits) {
+    switch (bits) {
+        #if CONFIG_IDF_TARGET_ESP32
+        case 9:
+            self->width = ADC_WIDTH_BIT_9;
+            break;
+        case 10:
+            self->width = ADC_WIDTH_BIT_10;
+            break;
+        case 11:
+            self->width = ADC_WIDTH_BIT_11;
+            break;
+        #endif
+        #if CONFIG_IDF_TARGET_ESP32 || CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32S3
+        case 12:
+            self->width = ADC_WIDTH_BIT_12;
+            break;
+        #endif
         #if CONFIG_IDF_TARGET_ESP32S2
-        adc1_config_width(ADC_WIDTH_BIT_13);
-        #else
-        adc1_config_width(ADC_WIDTH_BIT_12);
+        case 13:
+            self->width = ADC_WIDTH_BIT_13;
+            break;
         #endif
-        adc_bit_width = 12;
-        initialized = 1;
+        default:
+            mp_raise_ValueError(MP_ERROR_TEXT("invalid bits"));
+    }
+    self->bits = bits;
+
+    if (self->unit_id == ADC_UNIT_1) {
+        adc1_config_width(self->width);
+    }
+    for (adc_atten_t atten = ADC_ATTEN_DB_0; atten < ADC_ATTEN_MAX; atten++) {
+        if (self->characteristics[atten] != NULL) {
+            esp_adc_cal_characterize(self->unit_id, atten, self->width, DEFAULT_VREF, self->characteristics[atten]);
```

Comment:
This looks very much like a calibration call, which could be exposed via `.calibrate()`.  But I guess it's no use to call it multiple times?

Reply:
It's just retrieving the parameters from eFuse and calculating the characteristics for this attenuation/bit-width from them, rather than running any hardware calibration process (falling back on a simpler curve using `DEFAULT_VREF` if no parameters are available for the package). It'll always return the same results for the same inputs.

Reply:
Ok.